### PR TITLE
p7zip: Add initial package based on 16.02 sources

### PR DESCRIPTION
**Pull request by jave27**
- Also adds CVE and build patches from conda-forge and Debian

- [7zip] already exists but is Windows-only. This is for all other platforms.

Specify library name and version:  **p7zip/16.02**

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
It's looking for g++ when using clang profile. Probably need to force CC and CXX environment variables or patch its build script.

**Comment by jave27** [comment]
> It's looking for g++ when using clang profile. Probably need to force CC and CXX environment variables or patch its build script.

Yep, just pushed a new version that picks the right compiler.

**Review by uilianries — COMMENTED** [review]
@jave27 please, consider https://github.com/jave27/conan-center-index/pull/1

Also, patch description need to be updated, the link should be added to `patches_source`, the description should should give an idea what is about, a short detail.

**Comment by jave27** [comment]
@uilianries Thanks, I updated patch_source and patch_description fields.

**Review by uilianries — DISMISSED** [review]
LGTM

File: recipes/p7zip/all/test_package/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from conan import ConanFile
+from conan.tools.build import can_run
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "VirtualRunEnv"
+
+    def test(self):
```
Comment (prince-chrismc) [inline recipes/p7zip/all/test_package/conanfile.py:9]: I believe these are needed for 2.0 client

```suggestion
    generators = "VirtualRunEnv"
    test_type = "explicit"

    def build_requirements(self):
        self.tool_requires(self.tested_reference_str)

    def test(self):
```

Comment (prince-chrismc) [reply recipes/p7zip/all/test_package/conanfile.py:9]: https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/prebuilt_tool_package/all/test_package/conanfile.py

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**